src/main/java/org/prebid/server/bidder/madvertise/MadvertiseBidder.java

```diff
@@ -0,0 +1,165 @@
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.of(extractBids(bidResponse), Collections.emptyList());
+        } catch (DecodeException | PreBidException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private ExtImpMadvertise parseImpExt(Imp imp) {
+        final ExtImpMadvertise extImpMadvertise;
+        final String impId = imp.getId();
+
+        try {
+            extImpMadvertise = mapper.mapper().convertValue(imp.getExt(), MADVERTISE_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(String.format("Missing bidder ext in impression with id: %s", impId));
+        }
+
+        if (extImpMadvertise == null) {
+            throw new PreBidException(String.format("Missing bidder ext in impression with id: %s", impId));
+        }
+
+        final String zoneId = extImpMadvertise.getZoneId();
+
+        if (zoneId == null || zoneId.length() < ZONE_ID_MIN_LENGTH) {
+            throw new PreBidException(String.format("The minLength of zone ID is 7; ImpID=%s", impId));
+        }
+        return extImpMadvertise;
+    }
+
+    private HttpRequest<BidRequest> createRequest(BidRequest request, String zoneID) {
+        final String url = endpointUrl.replace(ZONE_ID_MACRO, HttpUtil.encodeUrl(zoneID));
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(url)
+                .headers(resolveHeaders(request.getDevice()))
+                .payload(request)
+                .body(mapper.encode(request))
+                .build();
+    }
+
+    private static MultiMap resolveHeaders(Device device) {
+        final MultiMap headers = HttpUtil.headers();
+
+        headers.add(HttpUtil.X_OPENRTB_VERSION_HEADER, X_OPENRTB_VERSION);
+        if (device != null) {
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.USER_AGENT_HEADER, device.getUa());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, HttpUtil.X_FORWARDED_FOR_HEADER, device.getIp());
+        }
+
+        return headers;
+    }
+
+    private static List<BidderBid> extractBids(BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .filter(Objects::nonNull)
+                .map(bid -> BidderBid.of(bid, getBidMediaType(bid), bidResponse.getCur()))
+                .collect(Collectors.toList());
+    }
+
+    private static BidType getBidMediaType(Bid bid) {
+        final List<Integer> videoBidAttrs = Arrays.asList(16, 6, 7);
```

Comment:
It's should be set, that can be created with `ImmutableSet.of(16, 6, 7);` and also this should be on class level